### ISSUE TITLE
Duplicate URLs on multilanguage, force `locale` on URL

### ISSUE THREAD

**Issue by doup** [bolt/core#2160]
Different URLs load the same content (bad for SEO).

Details
-------

| Question                                              | Answer
|------------------------------------------------------ | ------
| Relevant Bolt Version                                 | 4.1
| Install type                                          |  Composer install
| <abbr title="Backwards Compatibility">BC</abbr> Break | yes
| PHP version                                           | 7.4 |
| Web server                                            | Built-in
| For UX/UI issues                                      | –

Reproduction
------------

### Bug summary

- When multiple locales are enabled, the non localized URL and the default locale URL show the same content, e.g:
  - `/projects`
  - `/ca/projects`

### Steps to reproduce

- Create `contenttype` with locales:
    ```
    projects:
        name: Projects
        singular_name: Project
        locales: [ca, es]
    ```
- Set a locale in `services.yml`:
    ```
    parameters:
        locale: ca
    ```
- The following URLs show the same page:
  - `/projects`
  - `/ca/projects`

### Expected result

- We should redirect all non-localized URLs to their localized counterpart:
  - `/projects` => `/ca/projects`
- An option to force the locale on **all** URLs would be handy, so even if the contenttype doesn't define a `locales` array it gets redirected to the default locale one, e.g.:
  - `/` => `/ca/`

### Actual result

- Both pages render correctly

**Comment by I-Valchev** [comment]
Hi @doup, first thanks for your detailed issue descriptions! Really a great example of how it should be done so people can review fast and come back to you in time!

I wonder if this one is an issue, however. For all pages, Bolt knows there is one which is the `canonical`, and the rest are _not_. So, as long as your theme properly uses the canonical (you can check the default theme to make sure), I do not think this is an issue in terms of SEO. Am I making sense?

**Comment by doup** [comment]
On the URL duplication, I understand and agree that it's a non-issue if a `canonical` tag is added on the theme. I think there are two approaches here:

- Themes are responsible. I've double checked the docs and I suppose its just a matter of using `{{ canonical() }}` or `{{ record|link() }}` on my twig templates.
- Bolt is responsible. Make an opinionated choice to avoid this kind of duplication (related #2159). In the list of URL definitions/matchers that Bolt provides by default there should be only one way to access a certain record or list. So, for a combination of page type (list/record), content type & language there should be just one URL. Now there are up to four:
  - `/projects`
  - `/project`
  - `/ca/projects`
  - `/ca/project`

Then, there is the issue of prefixing the default locale or not, I think it should be at least configurable. So that `/` redirects to the default locale (e.g. `/ca/`). If this would be an option, then on a website which is already on production it's trivial to change the default locale, deprecate locale, etc. otherwise it might be a mess. I mention this because Bolt uses the non prefixed URL as canonical for the default language.

Thank you! ^_^

**Comment by doup** [comment]
I've reviewed how canonical URLs are working and I found a couple of things:

- **Listings** canonical for the default locale is broken, it just "echos" the URL. This is only a problem with the default locale as there are two valid URLs, e.g.: `/en/pages` & `/pages`.
  - To check this: add `{{canonical()}}` on a listing template and add/remove the default locale prefix in the URL
- **Records** canonical
  - Always uses the singular slug, [see `ContentHelper::getCanonicalRouteAndParams()`](https://github.com/bolt/core/blob/4.1/src/Utils/ContentHelper.php#L82)
  - On the default locale, it forces the URL without the locale, [see `Canonical::generateLink()`](https://github.com/bolt/core/blob/4.1/src/Canonical.php#L167-L177)

**I've a proposal:**

- Add a new config option to tell Bolt if the canonical URL should force-prefix the default locale or not
  - This could be interesting as you might start with a single locale (`en`), tell it to force the URLs without the default locale (`/pages`). Later on you decide to add a new locale (`es`), you could leave it as it was (e.g. `/pages` & `/es/pages`), but if you prefer to have for consistency all the URLs prefixed you could change this config so the canonical URLs always have the locale (e.g. `/en/pages` & `/es/pages`).
- Fix listings canonical for default locale, based on the previous config flag
- Use plural slug for the Records to make it consistent with Listings
  - Similar to what you would do on a REST API, e.g. `/pages` & `/pages/:id`
- Force redirect to the canonical URL.
  - If the user enters `/fr/page/slug` it gets redirected to `/fr/pages/slug`

What do you think?

**Comment by bobdenotter** [comment]
Hi, @doup 

> Different URLs load the same content (bad for SEO).

This is absolutely not true. If those different URLS have the same canonical, there's no issue whatsoever. 

About your suggestions: First two are fine with me! yes, let's get those fixed: 

these two however, i don't agree with: 

> Use plural slug for the Records to make it consistent with Listings

I don't like that. In my opionion `page/contact` looks nicer than `pages/contact`. It's a _single_ page, in the end.

If you _really_ don't want this, you can always set `sungular_slug` to the same as `slug` in `contenttypes.yaml`. 

> Force redirect to the canonical URL.

Well, we could do that, but I'm not sure there's any benefit to doing so. Do you have any links to corroborate that this will have any benefit on ranking of your site?

**Comment by doup** [comment]
> Different URLs load the same content (bad for SEO).
>> On the URL duplication, I understand and agree that it's a non-issue if a canonical tag is added on the theme.

I see your point, and I already agreed ☝️. Actually, I just noticed about a misunderstanding though… previously I searched on the templates if `rel="canonical"` was used and found nothing. I though that this is something you expected the theme creators to take care of, but I just realized that `rel="canonical"` is injected by Bolt automatically. **So there is no SEO issue**.

---

1. Created: https://github.com/bolt/core/issues/2219
2. I double checked before opening a new issue and I think I did something wrong, for listing pages it always shows the non-localized URL. 🤦  But, I found this issue: https://github.com/bolt/core/issues/2218
3. I can live with setting the plural form in the `singular_slug` 👍 
4. Yes, SEO wise, force redirecting doesn't add anything if the canonical tag is already injected. Actually is an alternative way of enforcing canonical where no tag injection would be needed.
